### PR TITLE
feat: implement retry option on message failure to send [FS-1571]

### PR DESCRIPTION
**Pull request by V-Gira**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/FS-1571" title="FS-1571" target="_blank">Sub-taskFS-1571</a>  [web] implement retry mechanism for messages that have completely failed to send
  </td></table>
  <br />
 

### Feature

- implement an option to retry sending a message if sending failed
- implementation with this PR only works with text messages, asset messages display a discard button instead
Peek 2023-03-07 12-18

### CODE REVIEW

File: src/script/conversation/MessageRepository.ts [hunk 1]
```diff
@@ -327,13 +327,14 @@ export class MessageRepository {
...
     quoteEntity?: OutgoingQuote,
+    messageId?: string,
   ): Promise<void> {
     const textPayload = {
       conversation,
       mentions,
       message: textMessage,
-      messageId: createRandomUuid(), // We set the id explicitely in order to be able to override the message if we generate a link preview
       quote: quoteEntity,
+      messageId: messageId ? messageId : createRandomUuid(), // We set the id explicitely in order to be able to override the message if we generate a link preview
```
Comment (atomrc) [inline src/script/conversation/MessageRepository.ts]:
```suggestion
      messageId: messageId ?? createRandomUuid(), // We set the id explicitely in order to be able to override the message if we generate a link preview
```

Comment (V-Gira) [reply src/script/conversation/MessageRepository.ts]: tried to put a more appropriate comment too

File: src/script/components/MessagesList/Message/ContentMessage/ContentMessage.tsx [hunk 2]
```diff
@@ -192,6 +200,34 @@ const ContentMessageComponent: React.FC<ContentMessageProps> = ({
...
+        incomingQuote && isOutgoingQuote(incomingQuote) ? (incomingQuote as OutgoingQuote) : undefined;
+
+      await messageRepository.sendTextWithLinkPreview(conversation, messageText, mentions, quote, messageId);
+    }
+  };
+
+  const handleDiscard = async () => {
+    await messageRepository.deleteMessageById(conversation, message.id);
+  };
+
```
Comment (atomrc) [inline src/script/components/MessagesList/Message/ContentMessage/ContentMessage.tsx]: This logic doesn't really belong there. 
The fact that you needed to inject the `messageRepository` in this component is an indicator that the logic might not be in the correct layer of the app. 

If you look at the codebase, you will see that the `MessageWrapper` already has access to the `MessageRepository`, so it should be to one implementing `retry` and `discard`. 

This component should have 2 callbacks as props `onRetry` and `onDiscard` and this logic should live in the `MessageWrapper`

File: src/script/conversation/MessageRepository.ts [hunk 3]
```diff
@@ -327,13 +327,16 @@ export class MessageRepository {
...
+    messageId?: string,
   ): Promise<void> {
     const textPayload = {
       conversation,
       mentions,
       message: textMessage,
-      messageId: createRandomUuid(), // We set the id explicitely in order to be able to override the message if we generate a link preview
       quote: quoteEntity,
+      // We set the id explicitely in order to be able to override the message if we generate a link preview
+      // Similarly, we provide that same id when we retry to send a failed message in order to override the original
```
Comment (atomrc) [inline src/script/conversation/MessageRepository.ts:338]: 👍

**Review by atomrc — CHANGES_REQUESTED** [review]
Super solid PR, Just one thing that is slightly misplace but overall great job @V-Gira

File: src/script/components/MessagesList/Message/ContentMessage/ContentMessage.tsx [hunk 4]
```diff
@@ -234,7 +241,21 @@ const ContentMessageComponent: React.FC<ContentMessageProps> = ({
...
+        {failedToSend && (
+          <PartialFailureToSendWarning failedToSend={failedToSend} knownUsers={conversation.allUserEntities} />
+        )}
+
+        {status === StatusType.FAILED && (
+          <CompleteFailureToSendWarning
+            isTextAsset={message.getFirstAsset().isText()}
+            onDiscard={() => {
+              onDiscard();
+            }}
```
Comment (atomrc) [inline src/script/components/MessagesList/Message/ContentMessage/ContentMessage.tsx]:
```suggestion
            onDiscard={onDiscard}
```

WDYT?

File: src/script/components/MessagesList/Message/ContentMessage/ContentMessage.tsx [hunk 5]
```diff
@@ -234,7 +241,21 @@ const ContentMessageComponent: React.FC<ContentMessageProps> = ({
...
+            onRetry={() => {
+              onRetry(message);
+            }}
```
Comment (atomrc) [inline src/script/components/MessagesList/Message/ContentMessage/ContentMessage.tsx]:
```suggestion
            onRetry={() => onRetry(message)}
```
WDYT?

**Review by atomrc — APPROVED** [review]
😎  Looking great. Good job man 👏

File: src/script/components/MessagesList/Message/ContentMessage/ContentMessage.tsx (hunk 5 above)
Comment (V-Gira) [reply src/script/components/MessagesList/Message/ContentMessage/ContentMessage.tsx]: forgot to remove dem curly braces, good catch

File: src/script/components/MessagesList/Message/MessageWrapper.tsx [hunk 6]
```diff
@@ -90,6 +95,29 @@ export const MessageWrapper: React.FC<MessageParams & {hasMarker: boolean; isMes
     }
   };
 
+  const onRetry = async (message: ContentMessage) => {
+    const firstAsset = message.getFirstAsset();
+
+    if (firstAsset instanceof Text) {
+      const messageId = message.id;
+      const messageText = firstAsset.text;
+      const mentions: MentionEntity[] = firstAsset.mentions() && firstAsset.mentions();
```
Comment (przemvs) [inline src/script/components/MessagesList/Message/MessageWrapper.tsx:104]: Hmm ?
```suggestion
      const mentions: MentionEntity[] = firstAsset?.mentions?.() || [];
```

File: src/script/components/MessagesList/Message/MessageWrapper.tsx [hunk 7]
```diff
@@ -90,6 +95,29 @@ export const MessageWrapper: React.FC<MessageParams & {hasMarker: boolean; isMes
...
+      const incomingQuote = message.quote();
+
+      const isOutgoingQuote = (quoteEntity: QuoteEntity): quoteEntity is OutgoingQuote => {
+        return quoteEntity.hash !== undefined;
+      };
```
Comment (przemvs) [inline src/script/components/MessagesList/Message/MessageWrapper.tsx:109]: Maybe move this to Guards folder or above the component :)

File: src/script/conversation/EventMapper.ts [hunk 8]
```diff
@@ -145,8 +145,8 @@ export class EventMapper {
     const {id, data: eventData, edited_time: editedTime, conversation, qualified_conversation} = event;
 
     if (eventData.quote) {
-      const {message_id: messageId, user_id: userId, error} = eventData.quote;
-      originalEntity.quote(new QuoteEntity({error, messageId, userId}));
+      const {hash: hash, message_id: messageId, user_id: userId, error} = eventData.quote;
```
Comment (przemvs) [inline src/script/conversation/EventMapper.ts:148]:
```suggestion
      const {hash, message_id: messageId, user_id: userId, error} = eventData.quote;
```

File: src/script/conversation/MessageRepository.ts [hunk 9]
```diff
@@ -1132,6 +1140,20 @@ export class MessageRepository {
     return undefined;
   }
 
+  private async updateMessageAsFailed(conversationEntity: Conversation, eventId: string) {
+    try {
+      const messageEntity = await this.getMessageInConversationById(conversationEntity, eventId);
+      const updatedStatus = StatusType.FAILED;
```
Comment (przemvs) [inline src/script/conversation/MessageRepository.ts:1146]: Move it outside try {} catch()

**Pull request merged**